src/rules.c

```diff
@@ -0,0 +1,185 @@
+#include "rules.h"
+#include "aggregate/expr/expression.h"
+#include "spec.h"
+
+arrayof(SchemaRule*) SchemaRules_g;
+TrieMap *ScemaPrefixes_g;
+
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+SchemaRule *SchemaRule_Create(SchemaRuleArgs *args, IndexSpec *spec, QueryError *status) {
+  SchemaRule *rule = rm_calloc(1, sizeof(*rule));
+
+  rule->type = rm_strdup(args->type);
+  rule->filter_exp_str = args->filter_exp_str ? rm_strdup(args->filter_exp_str) : NULL;
+  rule->lang_field = args->lang_field ? rm_strdup(args->lang_field) : NULL;
+  rule->score_field = args->score_field ? rm_strdup(args->score_field) : NULL;
+  rule->payload_field = args->payload_field ? rm_strdup(args->payload_field) : NULL;
+
+  for (int i = 0; i < args->nprefixes; ++i) {
+    const char *p = rm_strdup(args->prefixes[i]);
+    rule->prefixes = array_ensure_append(rule->prefixes, &p, 1, const char*);
+  }
+
+  rule->spec = spec;
+
+  if (rule->filter_exp_str) {
+    rule->filter_exp = ExprAST_Parse(rule->filter_exp_str, strlen(rule->filter_exp_str), status);
+    if (!rule->filter_exp) {
+      QueryError_SetError(status, QUERY_EADDARGS, "Invalid expression");
+      goto error;
+    }
+  }
+
+  for (int i = 0; i < array_len(rule->prefixes); ++i) {
+    SchemaPrefixes_Add(rule->prefixes[i], spec);
+  }
+
+  SchemaRules_g = array_ensure_append_1(SchemaRules_g, rule);
```

Comment:
I think we do not need this anymore as we said that all the schemas are with rules

Reply:
And also, general comment, why not use array_append?

Reply:
@MeirShpilraien  It is a nice sugarcoat for array_append which also require `&`

Reply:
I do not understand why not just using array_append here?